frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboardView.tsx

```diff
@@ -0,0 +1,242 @@
+import type { UiParameter } from "metabase-lib/v1/parameters/types";
+import type {
+  Dashboard,
+  DashboardCard,
+  DashboardId,
+  ParameterValueOrArray,
+} from "metabase-types/api";
+import type { SelectedTabId } from "metabase-types/store";
+
+import { EmbedFrame } from "../../components/EmbedFrame";
+
+import { DashboardContainer } from "./PublicOrEmbeddedDashboard.styled";
+
+export function PublicOrEmbeddedDashboardView({
+  dashboard,
+  hasNightModeToggle,
+  isFullscreen,
+  isNightMode,
+  onFullscreenChange,
+  onNightModeChange,
+  onRefreshPeriodChange,
+  refreshPeriod,
+  setRefreshElapsedHook,
+  selectedTabId,
+  parameters,
+  parameterValues,
+  draftParameterValues,
+  setParameterValue,
+  setParameterValueToDefault,
+  dashboardId,
+  bordered,
+  titled,
+  theme,
+  hideParameters,
+  hideDownloadButton,
+  navigateToNewCardFromDashboard,
+  slowCards,
+  cardTitled,
+}: {
+  dashboard: Dashboard | null;
+  hasNightModeToggle?: boolean;
+  isFullscreen: boolean;
+  isNightMode: boolean;
+  onFullscreenChange: DashboardFullscreenControls["onFullscreenChange"];
+  onNightModeChange: EmbedThemeControls["onNightModeChange"];
+  onRefreshPeriodChange: DashboardRefreshPeriodControls["onRefreshPeriodChange"];
+  refreshPeriod: RefreshPeriod;
+  setRefreshElapsedHook: DashboardRefreshPeriodControls["setRefreshElapsedHook"];
+  selectedTabId: SelectedTabId;
+  parameters: UiParameter[];
+  parameterValues: Record<string, ParameterValueOrArray>;
+  draftParameterValues: Record<string, ParameterValueOrArray | null>;
+  setParameterValue: HandleThunkActionCreator<
+    typeof setParameterValueDashboardAction
+  >;
+  setParameterValueToDefault: HandleThunkActionCreator<
+    typeof setParameterValueToDefaultDashboardAction
+  >;
+  dashboardId: DashboardId;
+  bordered: boolean;
+  titled: boolean;
+  theme: DisplayTheme;
+  hideParameters: EmbedHideParameters;
+  hideDownloadButton: EmbedHideDownloadButton;
+  navigateToNewCardFromDashboard?: (
+    opts: NavigateToNewCardFromDashboardOpts,
+  ) => void;
+  slowCards: Record<number, boolean>;
+  cardTitled: boolean;
+}) {
+  const buttons = !isWithinIframe()
+    ? getDashboardActions({
+        dashboard,
+        hasNightModeToggle,
+        isFullscreen,
+        isNightMode,
+        onFullscreenChange,
+        onNightModeChange,
+        onRefreshPeriodChange,
+        refreshPeriod,
+        setRefreshElapsedHook,
+        isPublic: true,
+      })
+    : [];
+
+  const visibleDashcards = (dashboard?.dashcards ?? []).filter(
+    dashcard => !isActionDashCard(dashcard),
+  );
+
+  const dashboardHasCards = dashboard && visibleDashcards.length > 0;
+
+  const tabHasCards =
+    visibleDashcards.filter(
+      (dc: DashboardCard) => dc.dashboard_tab_id === selectedTabId,
+    ).length > 0;
+
+  const hiddenParameterSlugs = getHiddenParameterSlugs({
+    parameters,
+    dashboard,
+    selectedTabId,
+  });
```

Comment:
if these are potentially compute-intensive in a dashboard with a lot of dashcards, we could consider memoizing them. If not this should be fine.

Reply:
We are already doing this at the moment as we call `this.getHiddenParameterSlugs` on every render cycle.

https://github.com/metabase/metabase/blob/b228f16bb65795983031c648765192f1fa0e7eb8/frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx#L248

Look at its implementation, they're identical, I just extracted the dependency from `this.props` as props.

https://github.com/metabase/metabase/blob/b228f16bb65795983031c648765192f1fa0e7eb8/frontend/src/metabase/public/containers/PublicOrEmbeddedDashboard/PublicOrEmbeddedDashboard.tsx#L176-L188

It's true that I could have improved this, but I'd rather do that in a different PR maybe. This is a refactor PR, that shouldn't change the behavior whatsoever, for simplicity.

But thanks for the concern!